test/lib/allocator_test.rb

```diff
@@ -0,0 +1,97 @@
+# frozen_string_literal: true
+require_relative '../test_helper'
+
+describe GlobalUid::Allocator do
+  let(:connection)   { mock('connection') }
+  let(:increment_by) { 10 }
+  let(:allocator)    { GlobalUid::Allocator.new(incrementing_by: increment_by, connection: connection) }
+  let(:table)        { mock('table_class').class }
+
+  before do
+    restore_defaults!
+    connection.stubs(:current_database).returns('database_name')
+  end
+
+  describe 'when the auto_increment_increment differs to configuration' do
+    it 'prevents the connection to avoid allocating bad IDs' do
+      connection.stubs(:select_value).with('SELECT @@auto_increment_increment').returns(50)
+      exception = assert_raises(GlobalUid::InvalidIncrementException) do
+        GlobalUid::Allocator.new(incrementing_by: 10, connection: connection)
+      end
+
+      assert_equal("Configured: '10', Found: '50' on 'database_name'", exception.message)
+    end
+  end
+
+
+  describe 'with a configured increment_by matching the server auto_increment_increment' do
+    before do
+      connection.stubs(:select_value).with('SELECT @@auto_increment_increment').returns(increment_by)
+    end
+
+    describe '#allocate_one' do
+      it 'allocates IDs, maintaining a small rolling selection of IDs for comparison' do
+        [10, 20, 30, 40, 50, 60, 70, 80].each do |id|
+          connection.expects(:insert).returns(id)
+          allocator.allocate_one(table)
+        end
+
+        assert_equal(5, allocator.max_window_size)
+        assert_equal([40, 50, 60, 70, 80], allocator.allocations)
+      end
+
+      describe 'gap between ID not divisible by increment_by' do
+        it 'raises an error' do
+          connection.expects(:insert).returns(20)
+          allocator.allocate_one(table)
+
+          connection.stubs(:select_value).with('SELECT @@auto_increment_increment').returns(5)
+          connection.expects(:insert).returns(25)
+          exception = assert_raises(GlobalUid::InvalidIncrementException) do
+            allocator.allocate_one(table)
+          end
+
+          assert_equal("Configured: '10', Found: '5' on 'database_name'. Recently allocated IDs: [20, 25]", exception.message)
```

Comment:
I think we can make the exception message a bit easier to understand. How about e.g.

> The generated ID '25' is '5' larger than the last generated id, instead of the configured '10'. Recently allocated IDs: [20, 25].

Reply:
Probably that means we need different messages for negative increment and non-mod-zero increment. Wdyt?

Reply:
Considering how unlikely the exception is, I don't think it's worth the effort or added complexity. It's not as nice as the example you've provided but it's still informative.